driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -638,4 +709,277 @@ public Prune shouldPrune(final UsageTrackingInternalConnection usageTrackingConn
             return DefaultConnectionPool.this.shouldPrune(usageTrackingConnection) ? Prune.YES : Prune.NO;
         }
     }
+
+    /**
+     * Package-private methods are thread-safe,
+     * and only they should be called outside of the {@link OpenConcurrencyLimiter}'s code.
+     */
+    @ThreadSafe
+    private final class OpenConcurrencyLimiter {
```

Comment:
The spec tests are a pretty indirect test of this class.  Can you think of a way to unit test it that would provide any more coverage of edge cases?

Reply:
Thank you, this request proved to be extremely helpful: by virtue of the new tests I found and fixed two concurrency issues (both race conditions, not data races) in this PR.

The added tests check not only overall concurrent usage of the `DefaultConnectionPool`, but also the hand-over mechanism. There are, however, no tests for checkout/hand-over fairness. I believe I can imagine how to write those, but I am not sure the value of such tests would be worth time invested.